diff --git a/src/amd/vulkan/radv_physical_device.c b/src/amd/vulkan/radv_physical_device.c
index 2c1540f..8794ff8 100644
--- a/src/amd/vulkan/radv_physical_device.c
+++ b/src/amd/vulkan/radv_physical_device.c
@@ -65,14 +65,23 @@ radv_perf_query_supported(const struct radv_physical_device *pdev)
 }
 
 static bool
-radv_taskmesh_enabled(const struct radv_physical_device *pdev)
+radv_mesh_enabled(const struct radv_physical_device *pdev)
 {
    const struct radv_instance *instance = radv_physical_device_instance(pdev);
 
    if (instance->debug_flags & RADV_DEBUG_NO_MESH_SHADER)
       return false;
 
-   return pdev->use_ngg && !pdev->use_llvm && pdev->info.gfx_level >= GFX10_3 && radv_compute_queue_enabled(pdev);
+   return pdev->use_ngg && !pdev->use_llvm &&
+          (pdev->info.gfx_level >= GFX10_3 || pdev->info.has_gfx1013_mesh_shading);
+}
+
+static bool
+radv_task_enabled(const struct radv_physical_device *pdev)
+{
+   return radv_mesh_enabled(pdev) &&
+          (pdev->info.gfx_level >= GFX10_3 || pdev->info.has_gfx1013_task_shading) &&
+          radv_compute_queue_enabled(pdev);
 }
 
 bool
@@ -897,7 +906,7 @@ radv_physical_device_get_supported_extensions(const struct radv_physical_device
       .EXT_map_memory_placed = true,
       .EXT_memory_budget = true,
       .EXT_memory_priority = true,
-      .EXT_mesh_shader = radv_taskmesh_enabled(pdev),
+      .EXT_mesh_shader = radv_mesh_enabled(pdev),
       .EXT_multi_draw = true,
       .EXT_multisampled_render_to_single_sampled = radv_msrtss_enabled(pdev),
       .EXT_mutable_descriptor_type = true, /* Trivial promotion from VALVE. */
@@ -998,7 +1007,9 @@ static void
 radv_physical_device_get_features(const struct radv_physical_device *pdev, struct vk_features *features)
 {
    const struct radv_instance *instance = radv_physical_device_instance(pdev);
-   bool taskmesh_en = radv_taskmesh_enabled(pdev);
+   bool mesh_en = radv_mesh_enabled(pdev);
+   bool task_en = radv_task_enabled(pdev);
+   bool full_mesh_features = mesh_en && pdev->info.gfx_level >= GFX10_3;
    bool has_perf_query = radv_perf_query_supported(pdev);
    bool has_shader_image_float_minmax = pdev->info.gfx_level <= GFX7 || pdev->info.gfx_level == GFX10 ||
                                         pdev->info.gfx_level == GFX10_3 || pdev->info.gfx_level >= GFX12;
@@ -1354,10 +1365,10 @@ radv_physical_device_get_features(const struct radv_physical_device *pdev, struc
       .minLod = true,
 
       /* VK_EXT_mesh_shader */
-      .meshShader = taskmesh_en,
-      .taskShader = taskmesh_en,
-      .multiviewMeshShader = taskmesh_en,
-      .primitiveFragmentShadingRateMeshShader = taskmesh_en,
+      .meshShader = mesh_en,
+      .taskShader = task_en,
+      .multiviewMeshShader = full_mesh_features,
+      .primitiveFragmentShadingRateMeshShader = full_mesh_features,
       .meshShaderQueries = pdev->info.has_gfx1013_mesh_queries,
 
       /* VK_EXT_depth_clip_control */
@@ -1787,8 +1798,9 @@ radv_get_physical_device_properties(struct radv_physical_device *pdev)
    const bool has_fp16 = radv_shader_fp16_enabled(pdev);
    const bool enable_sparse = radv_sparse_enabled(pdev);
 
-   VkShaderStageFlags taskmesh_stages =
-      radv_taskmesh_enabled(pdev) ? VK_SHADER_STAGE_MESH_BIT_EXT | VK_SHADER_STAGE_TASK_BIT_EXT : 0;
+   VkShaderStageFlags taskmesh_stages = radv_mesh_enabled(pdev) ? VK_SHADER_STAGE_MESH_BIT_EXT : 0;
+   if (radv_task_enabled(pdev))
+      taskmesh_stages |= VK_SHADER_STAGE_TASK_BIT_EXT;
    VkShaderStageFlags rt_stages = radv_enable_rt(pdev) ? RADV_RT_STAGE_BITS : 0;
 
    bool accel_dot = pdev->info.compiler_info.has_accelerated_dot_product;
@@ -2344,7 +2355,7 @@ radv_get_physical_device_properties(struct radv_physical_device *pdev)
       .pipelineBinaryCompressedData = false,
 
       /* VK_KHR_compute_shader_derivatives */
-      .meshAndTaskShaderDerivatives = radv_taskmesh_enabled(pdev),
+      .meshAndTaskShaderDerivatives = radv_task_enabled(pdev),
 
       /* VK_EXT_device_generated_commands */
       .maxIndirectPipelineCount = 4096,
@@ -2743,7 +2755,7 @@ radv_physical_device_try_create(struct radv_instance *instance, drmDevicePtr drm
 
    pdev->emulate_ngg_gs_query_pipeline_stat = pdev->use_ngg && pdev->info.gfx_level < GFX11;
 
-   pdev->emulate_mesh_shader_queries = pdev->info.gfx_level == GFX10_3;
+   pdev->emulate_mesh_shader_queries = pdev->info.gfx_level == GFX10_3 || pdev->info.has_gfx1013_mesh_queries;
 
    /* PKT3_LOAD_SH_REG_INDEX is supported on GFX8+, but it hangs with compute queues until GFX10.3. */
    pdev->load_grid_size_from_user_sgpr = pdev->info.gfx_level >= GFX10_3;
diff --git a/src/amd/vulkan/radv_pipeline.c b/src/amd/vulkan/radv_pipeline.c
index 6e48fa1..838b088 100644
--- a/src/amd/vulkan/radv_pipeline.c
+++ b/src/amd/vulkan/radv_pipeline.c
@@ -517,7 +517,9 @@ radv_postprocess_nir(const struct radv_compiler_info *compiler_info, const struc
          .wave_size = stage->info.wave_size,
          .workgroup_size = stage->info.workgroup_size,
          .use_llvm = use_llvm,
-         .load_grid_size_from_user_sgpr = compiler_info->key.load_grid_size_from_user_sgpr,
+         .load_grid_size_from_user_sgpr =
+            compiler_info->key.load_grid_size_from_user_sgpr ||
+            (stage->stage == MESA_SHADER_TASK && compiler_info->key.has_gfx1013_task_shading),
       });
    NIR_PASS(_, stage->nir, radv_nir_lower_abi, gfx_level, stage, gfx_state, compiler_info->hw.address32_hi);
 
diff --git a/src/amd/vulkan/radv_query.c b/src/amd/vulkan/radv_query.c
index bf1ca5b..c7cea5b 100644
--- a/src/amd/vulkan/radv_query.c
+++ b/src/amd/vulkan/radv_query.c
@@ -411,7 +411,8 @@ radv_get_pipelinestat_query_size(struct radv_device *device)
      * invocations, it's easier to use the same size as GFX11.
      */
    const struct radv_physical_device *pdev = radv_device_physical(device);
-   unsigned num_results = pdev->info.gfx_level >= GFX10_3 ? 14 : 11;
+   unsigned num_results =
+      pdev->info.gfx_level >= GFX10_3 || pdev->info.has_gfx1013_mesh_queries ? 14 : 11;
    return num_results * 8;
 }
diff --git a/src/amd/vulkan/radv_queue.c b/src/amd/vulkan/radv_queue.c
index 93f71ec..1f22dd6 100644
--- a/src/amd/vulkan/radv_queue.c
+++ b/src/amd/vulkan/radv_queue.c
@@ -912,7 +912,7 @@ radv_update_preamble_cs(struct radv_queue_state *queue, struct radv_device *devi
    }
 
    if (!queue->ring_info.task_rings && needs->task_rings) {
-      assert(pdev->info.gfx_level >= GFX10_3);
+      assert(pdev->info.gfx_level >= GFX10_3 || pdev->info.has_gfx1013_task_shading);
 
       /* We write the control buffer from the CPU, so need to grant CPU access to the BO.
        * The draw ring needs to be zero-initialized otherwise the ready bits will be incorrect.
@@ -932,7 +932,7)
    }
 
    if (!queue->ring_info.mesh_scratch_ring && needs->mesh_scratch_ring) {
-      assert(pdev->info.gfx_level >= GFX10_3);
+      assert(pdev->info.gfx_level >= GFX10_3 || pdev->info.has_gfx1013_task_shading);
       result =
          radv_bo_create(device, NULL, AC_MESH_SCRATCH_NUM_ENTRIES * AC_MESH_SCRATCH_ENTRY_BYTES, 256,
                         RADEON_DOMAIN_VRAM, ring_bo_flags, RADV_BO_PRIORITY_SCRATCH, 0, true, &mesh_scratch_ring_bo);
diff --git a/src/amd/vulkan/radv_shader.c b/src/amd/vulkan/radv_shader.c
index 4867cb2..2a5f04a 100644
--- a/src/amd/vulkan/radv_shader.c
+++ b/src/amd/vulkan/radv_shader.c
@@ -3420,6 +3420,8 @@ radv_aco_fill_compiler_options(struct aco_compiler_options *aco_info, const stru
    aco_info->gfx_level = compiler_info->ac->gfx_level;
    aco_info->family = compiler_info->debug.family;
    aco_info->address32_hi = compiler_info->hw.address32_hi;
+   aco_info->has_mesh_shading =
+      compiler_info->ac->gfx_level >= GFX10_3 || compiler_info->key.has_gfx1013_mesh_shading;
 }
 
 void
diff --git a/src/amd/vulkan/radv_shader.h b/src/amd/vulkan/radv_shader.h
index 7f7611a..931152c 100644
--- a/src/amd/vulkan/radv_shader.h
+++ b/src/amd/vulkan/radv_shader.h
@@ -539,6 +539,8 @@ struct radv_compiler_info {
       uint32_t nggc_max_ps_params : 4;
       uint32_t no_ngg_gs : 1;
       uint32_t load_grid_size_from_user_sgpr : 1;
+      uint32_t has_gfx1013_mesh_shading : 1;
+      uint32_t has_gfx1013_task_shading : 1;
       uint32_t emulate_ngg_gs_query_pipeline_stat = pdev->emulate_ngg_gs_query_pipeline_stat,
       uint32_t primitives_generated_query : 1;
       uint32_t mesh_shader_queries : 1;
@@ -566,7 +568,7 @@ struct radv_compiler_info {
       uint32_t no_implicit_varying_subgroup_size : 1;
       uint32_t force_nan_preserve_min_max : 1;
       uint32_t nir_debug_info : 1;
-      uint32_t padding : 28;
+      uint32_t padding : 26;
 
       int32_t force_aniso;
 
diff --git a/src/amd/vulkan/radv_shader_args.c b/src/amd/vulkan/radv_shader_args.c
index 8c36d19..c0d9f95 100644
--- a/src/amd/vulkan/radv_shader_args.c
+++ b/src/amd/vulkan/radv_shader_args.c
@@ -620,7 +620,8 @@ declare_shader_args(const struct radv_compiler_info *compiler_info, struct radv_
       declare_global_input_sgprs(state, gfx_level, info, user_sgpr_info);
 
       if (info->cs.uses_grid_size) {
-         if (compiler_info->key.load_grid_size_from_user_sgpr)
+         if (compiler_info->key.load_grid_size_from_user_sgpr ||
+             (stage == MESA_SHADER_TASK && compiler_info->key.has_gfx1013_task_shading))
             RADV_ADD_UD_ARG(state, 3, AC_ARG_VALUE, ac.num_work_groups, AC_UD_CS_GRID_SIZE);
          else
             RADV_ADD_UD_ARG(state, 2, AC_ARG_CONST_ADDR, ac.num_work_groups, AC_UD_CS_GRID_SIZE);
